#25 disabilities

loffleraSMH opened the pull request:
Adds a new function to derive disability flag.

Closes issue #25 

To test this function, you will need to re-install the package from the corresponding branch in order to successfully read the data file containing the diagnosis code mapping (e.g., `pak::pkg_install("GEMINI-Medicine/Rgemini@25_disabilities")`.

Note: The documentation still needs to be improved to provide a better reference/rationale for when/why users may want to exclude ER diagnosis codes (also see issues #43 and #44 - I've added a note in both issues that this also needs to be updated in `disability` as well so this could be tackled when working on those issues).

loffleraSMH commented:
Hey @gemini-wenb, I don't know if you already started reviewing this function, but I just pushed some more changes:

1. I liked your version of the `component_wise` output in `frailty_score` much better than my original implementation of it, so I changed it to be more similar to yours (with one small difference: I'm also including the code descriptions in the output, which I thought might be helpful for users to understand/check the mappings).
2. I made some small changes to the mapping file, so you'll likely have to re-install the branch for the function to work as expected.
3. I also added the input checks that were previously commented out because the `check_input` function has now been merged into `develop`.

gemini-wenb reviewed (commented):
This is a draft review focuses on larger/essential issues for developer to see higher-level comments in the interest of time efficiency. A more detailed review will followed.

loffleraSMH commented on tests/testthat/test-disability.R:
```diff
@@ -0,0 +1,74 @@
+  ## Check 2: Unit test for some diagnosis codes that should have disability = TRUE
+  check2 <- disability(cohort = data.table(genc_id = c(1,2,3,4,5)),
+                       ipdiag = data.table(genc_id = c(1,1,2,2,2,3,4,5,5,5,5,5),
+                                           diagnosis_code = c('F7001', 'Q8723','Q66', 'E2204', 'M23836', 'G60', 'G242','S029', 'Z998', 'H90132', 'Q150', 'E10352')),
```
I generally agree that we should use the dummy functions as much as possible. The issue I ran into was that some diagnoses I wanted to test weren't created with the dummy function. E.g., running this `ipdiag <- dummy_diag(nid=10, nrow=100, ipdiagnosis=T, pattern ="^F700")` results in an error, probably because F700 is a Canadian code, so I'd need to connect to the DB to create this code. It seemed easier to just make up fake examples of things I wanted to test than check whether all the different scenarios I wanted to test are covered by the randomly generated dummy data. However, if you feel strongly about this, I can pick other diagnosis prefixes and make sure the dummy data covers all the scenarios I want to test.

gemini-wenb replied:
Yes if you want to sample from Canadian code right now a db connection is needed - could be a future enhancement.  Generally I would think if a function picks up e.g. "^F840" it would work the same for F700. So if an alternative can achieve the same goal for the test I'd try to use the dummy fun. If there is a specific reason that we want to test specifically for F700 I don't feel strongly about diverging from using dummy fun.

loffleraSMH replied:
Yes, I think the main problem is that a lot of code prefixes only return the exact same code with the `dummy_diag` function because of that (e.g., "^F840" only returns exactly codes = "F840"), so I had to hand-pick a few other codes to make sure I'm also getting some that are actually children codes of the ones in the mapping file. This can easily be missed when using the dummy functions, so just something to keep in mind (plus, in this case, I wanted to cover all the different categories so had to generate an unnecessarily large dummy input to achieve that). Anyway, I did change the test as you suggested and I think it's fine now.

gemini-wenb replied:
All make sense! Pros and cons of sampling randomness. And sometimes the function doesn't return child level codes is due to that the child level codes do not actually exist (no further sublevels). We can keep this as a future enhancement - to allow greater user control on the codes returned and potentially allow sampling from icd-10-ca without db connection.

loffleraSMH commented on tests/testthat/test-disability.R:
```diff
@@ -0,0 +1,74 @@
+  expect_true(unique(check2[diagnosis_code %in% c("S029", "Z998"), disability_category]) == "Physical disability - Permanent Injuries")
+  expect_true(check2[diagnosis_code %in% c("H90132"), disability_category] == "Sensory disabilities - Hearing impairments")
+  expect_true(unique(check2[diagnosis_code %in% c("Q150", "E10352"), disability_category]) == "Sensory disabilities - Vision impairments")
+
```
Good point! Changed it for genc_id rows. I did keep separate tests for the categories because I changed the test slightly based on the updated version with `dummy_diag` and I think it's a bit easier to read that way.

gemini-wenb replied:
Thanks for making the updates! All looks good!

loffleraSMH commented on tests/testthat/test-disability.R:
```diff
@@ -0,0 +1,74 @@
+  check3 <- disability(cohort = data.table(genc_id = c(1,2,3,4,5,6,7,8,9,10,11,12)),
+                       ipdiag = data.table(genc_id = c(1,2,3,4,5,6,7,8,9,10,11,12),
+                                           diagnosis_code = c('I4891', 'A021','B962', 'K8308', 'Q68', 'G65', 'H91','F83', 'Z99', 'H92', 'E12', 'Z89')),
+                       erdiag = NULL, component_wise = TRUE)
```
Yep, in this case it was a bit easier to create the relevant dummy codes based on the parent pattern, so I changed this as suggested.

gemini-wenb commented on R/disability.R:
```diff
@@ -0,0 +1,181 @@
+#' The output is returned in long format, where each row corresponds to a
+#' disability diagnosis (`diagnosis_code`) and its corresponding
+#' `disability_category` (`character`).
+#'
```
The `@return` description is somewhat repetitive of what's described in `@param component_wise`.
I would shorten the description in `@param component_wise` by moving contents more about what is returned to the `return` section. It would be cleaner to have a more concise description for parameters.

gemini-wenb commented on R/disability.R:
```diff
@@ -0,0 +1,181 @@
+#' - Sensory disabilities - Vision impairments
+#' - Developmental Disabilities
+#'
+#' @section Warning:
```
Maybe call the section "Notes" instead of "Warning". For the frailty_score function, I dedicated warning sections for documenting important warning messages the function produces. Nothing crucial, just for consistency in terms of tag uses.

gemini-wenb reviewed (approved):
Thanks for developing this function and addressing comments in my previous review. Everything looks good - function works as expected with good documentation and unit tests. 
I only have 2 minor suggestions on the documentation, but happy to keep as it is.
Ready to merge!

loffleraSMH commented:
> Thanks for developing this function and addressing comments in my previous review. Everything looks good - function works as expected with good documentation and unit tests. I only have 2 minor suggestions on the documentation, but happy to keep as it is. Ready to merge!

Thanks for reviewing @gemini-wenb! I made the 2 small changes to the documentation you suggested and will go ahead and merge now.
